snapcraft/internal/repo/snaps.py

```diff
@@ -0,0 +1,215 @@
+        if self._original_channel:
+            snap_install_cmd.extend(['--channel', self._original_channel])
+        if self.is_classic:
+            # TODO make this a user explicit choice
+            snap_install_cmd.extend(['--classic'])
+        try:
+            check_call(snap_install_cmd)
+        except CalledProcessError as install_error:
+            raise errors.SnapInstallError(snap_name=self.name,
+                                          snap_channel=self.channel)
+
+    def refresh(self):
+        """Refreshes a snap onto a channel on the system."""
+        snap_refresh_cmd = []
+        if _snap_command_requires_sudo():
+            snap_refresh_cmd = ['sudo']
+        snap_refresh_cmd.extend(['snap', 'refresh', self.name,
+                                 '--channel', self.channel])
+        if self.is_classic:
+            # TODO make this a user explicit choice
+            snap_refresh_cmd.extend(['--classic'])
+        try:
+            check_call(snap_refresh_cmd)
+        except CalledProcessError as install_error:
+            raise errors.SnapRefreshError(snap_name=self.name,
+                                          snap_channel=self.channel)
+
+
+def install_snaps(snaps_list):
+    """Install snaps of the format <snap-name>/<channel>."""
+    for snap in snaps_list:
+        snap_pkg = SnapPackage(snap)
+        if not snap_pkg.installed and snap_pkg.in_store:
+            snap_pkg.install()
+        elif snap_pkg.get_current_channel() != snap_pkg.channel:
+            snap_pkg.refresh()
+
+
+def _snap_command_requires_sudo():
+    # snap whoami returns - if the user is not logged in.
+    output = check_output(['snap', 'whoami'])
+    whoami = output.decode(sys.getfilesystemencoding())
+    requires_root = False
+    try:
+        requires_root = whoami.split(':')[1].strip() == '-'
+    # A safeguard if the output changes
+    except IndexError:
+        requires_root = True
+    if requires_root:
+        logger.warning('snapd is not logged in, snap install '
+                       'commands will use sudo')
+    return requires_root
+
+
+def _get_parsed_snap(snap):
+    if '/' in snap:
+        sep_index = snap.find('/')
+        snap_name = snap[:sep_index]
+        snap_channel = snap[sep_index+1:]
+    else:
+        snap_name = snap
+        snap_channel = ''
+    return snap_name, snap_channel
+
+
+def _get_local_snap_info(snap_name):
+    slug = 'snaps/{}'.format(urllib.parse.quote(snap_name, safe=''))
+    url = _SNAPD_URI_ROOT_TEMPLATE.format(slug)
+    with requests_unixsocket.Session() as session:
+        snap_info = session.get(url)
+    snap_info.raise_for_status()
+    return snap_info.json()['result']
+
+
+def _get_store_snap_info(snap_name):
+    # This logic uses /v2/find returns an array of results, given that
+    # we do a strict search either 1 result or a 404 will be returned.
+    slug = '{}{}'.format('find?name=', urllib.parse.quote(snap_name, safe=''))
```

Comment:
huh? why are you interpolating a fixed string?

'find?name={}'.format(...)` would be more intuitive

Reply:
I don't know :upside_down_face:

Reply:
any reason for not using `urllib.parse.urlencode` here?

Reply:
FWIW using `urllib.parse.quote` like that results in unnecessary %-encoding of characters that don't need it. If you'd rather not use `urlencode` (maybe building a dict isn't to your taste?), just use `urllib.parse.quote_plus` which is the quote function for the query part of a URL.

Reply:
`requests_unixsockets` doesn't support the dict as params, but I guess you are implying that urlencode can do that so i'll take a look. The only one I sort of have problems with is `urljoin`.

Reply:
switched to `urlencode` on bec56ed